From 22ea4752ed4f9b83286e6ded5d9c290b25d73c26 Mon Sep 17 00:00:00 2001
From: Alexandru Ardelean <alexandru.ardelean@analog.com>
Date: Fri, 16 Nov 2018 09:50:28 +0200
Subject: [PATCH 19/25] dma: axi-dmac: terminate early DMA transfers after a
 partial one

When we get a partial transfer, we should not submit any more segments to
the hardware, as the data it will get will be unused and will just do
unnecessary processing.

During the start of transfer, we check it there was a partial transfer in a
previous completed segment and add the TLAST flag to the next segment that
is about to be submitted.

Signed-off-by: Alexandru Ardelean <alexandru.ardelean@analog.com>
(cherry picked from commit 70b6b10a3739e93ba3ae987459b972be7384cac2)
Signed-off-by: Zumeng Chen <zumeng.chen@windriver.com>
---
 drivers/dma/dma-axi-dmac.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/drivers/dma/dma-axi-dmac.c b/drivers/dma/dma-axi-dmac.c
index af2e111c7a33..384699b3d9b8 100644
--- a/drivers/dma/dma-axi-dmac.c
+++ b/drivers/dma/dma-axi-dmac.c
@@ -105,6 +105,7 @@ struct axi_dmac_sg {
 struct axi_dmac_desc {
 	struct virt_dma_desc vdesc;
 	bool cyclic;
+	bool have_partial_xfer;
 
 	unsigned int num_submitted;
 	unsigned int num_completed;
@@ -234,7 +235,8 @@ static void axi_dmac_start_transfer(struct axi_dmac_chan *chan)
 	}
 
 	desc->num_submitted++;
-	if (desc->num_submitted == desc->num_sgs) {
+	if (desc->num_submitted == desc->num_sgs ||
+	    desc->have_partial_xfer) {
 		if (desc->cyclic) {
 			desc->num_submitted = 0; /* Start again */
 		} else {
@@ -309,6 +311,7 @@ static void axi_dmac_dequeue_partial_xfers(struct axi_dmac_chan *chan)
 				if (sg->id == AXI_DMAC_SG_UNUSED)
 					continue;
 				if (sg->id == id) {
+					desc->have_partial_xfer = true;
 					sg->partial_len = len;
 					found_sg = true;
 					break;
-- 
2.17.0

